### PR TITLE
Allow users to reject their own watch/blacklist PRs

### PR DESCRIPTION
**Pull request by Spevacus**
Implements feature request #8642

This PR allows users to reject/close (including `-force`) their own auto-created watch and blacklist PRs.

This was tested in [this chatroom](https://chat.stackexchange.com/transcript/message/65699938#65699938) (where I removed SD ignoring its own messages when considering them for commands and some other changes, the full diff of which is available [here](https://github.com/Charcoal-SE/SmokeDetector/pull/11285/commits/d3993d17b832c66085ef56a5cd063cf09dd07ff1) where I mistakenly committed them to this branch and had to remove them).

One small change I made was making the `payload` argument for `get_pull_request` an optional arg with a blank string as its default. This change has no effect elsewhere in the project as `get_pull_request` isn't used anywhere aside from in this PR's changes, and you don't need to supply a payload when performing a GET request for a PR anyway.

I also took the liberty to remove the "so the user can learn from their mistakes" content from the required-rejection CmdException. I have a couple of reasons for this removal:
- This wording makes no sense if the user is rejecting their own PR
- We frequently reject duplicate PRs. In such a case, there is no mistake being made.

## Known limitations:
This PR operates off of the chat messager's chat id, and so does submitting a watch PR. This can be different across chat servers, such as if a user submits a watch PR in SO chat and attempts to self-close it in Charcoal HQ. This will yield a `"You need blacklist manager privileges to reject pull requests that aren't created by you."` error.

Given that 95% of all watch/blacklist PRs occur in Charcoal HQ, I consider this an acceptable limitation.

### CODE REVIEW

**Comment by Spevacus** [comment]
The current error is due to the GitHubManager not being spun-up during testing (or at least I think) - I'll resolve this at a later time.

**Comment by tripleee** [comment]
Thanks for this contribution; I'll take a closer look when I am less swamped.

File: chatcommands.py [hunk 1]
```diff
@@ -717,11 +717,21 @@ def reject(msg, args, alias_used="reject"):
         reason = ''
     force = alias_used.split("-")[-1] == "force"
     code_permissions = is_code_privileged(msg._client.host, msg.owner.id)
-    if not code_permissions:
-        raise CmdException("You need blacklist manager privileges to reject pull requests")
+    self_reject = False
+    try:
+        pr_json = GitHubManager.get_pull_request(pr_id).json()
+        if 'body' in pr_json:
+            pr_authored_by_rejector = regex.search(r"(?<=\/users\/)" + str(msg.owner.id),
```
[the remark this answers is not in the document]
Comment (tripleee) [reply chatcommands.py:724]: I guess the link actually contains the chat user ID for the creator of the PR, so that part should be fine.

**Review by tripleee — CHANGES_REQUESTED** [review]
On the whole, the code looks neat, but the commit history is really messy; can you rebase this to a more logical 4-5 commits max (realistically 1-3 I guess)?

File: gitmanager.py [hunk 2]
```diff
@@ -73,7 +73,7 @@ def comment_on_thread(cls, thread_id, body):
         return response.json()
 
     @classmethod
-    def get_pull_request(cls, pr_id, payload):
+    def get_pull_request(cls, pr_id, payload=""):
```
Comment (tripleee) [inline gitmanager.py:76]: Is this change necessary and useful?

Comment (tripleee) [reply gitmanager.py:76]: I guess this actually fixes a bug which had gone unnoticed?

Comment (Spevacus) [reply gitmanager.py:76]: Yeah, a payload isn't really required when sending a GET request broadly speaking. If you're supplying a payload, generally you'd be wanting to do so in a POST request. In this case, we just need to perform a GET to get a PR, so a blank payload is sufficient. This argument, really, could be completely removed.

File: chatcommands.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Spevacus) [reply chatcommands.py:724]: re: Commit history: Yeah, the commit history is horrendous, I agree. I'll clean that up.

`get_pull_request` returns a request response in string format. I call `.json()` on that response here to just convert it to json, then check for a `'body'` in it. If it doesn't exist in the json'd response, the request failed or had another issue, and I just let error handling after this account for that (as messing with our unit testing would be a scope change I'm not comfortable venturing into).

If the chat user ID is different from the user ID in the PR body (generated by the SmokeDetector git user) then it'll just not match here. Further, this won't affect PRs that weren't opened by the SD git user by virtue of the a

For the regex thing, yeah, I think I can drop the backslashes.

**Comment by makyen** [comment]
Personally, I don't have a problem with there being lots of commits, as that's sometimes useful in order to group changes into conceptual blocks, although in my own work, I do *tend* to go through and tidy commits into conceptual changes when possible/reasonable.

I do, however, have a problem with commit messages which are something like just "Update chatcommands.py". Such a commit message is counterproductive, because it gives almost no information as to what the changes are in the commit. Commit messages are part of what guides both reviewers and people using the code in the future when looking through what changes were made and what to take a more detailed look at. Please see #3 in [CONTRIBUTING_PULLREQS.md](https://github.com/Charcoal-SE/SmokeDetector/blob/master/CONTRIBUTING_PULLREQS.md).

I, generally, allow such commit messages to slip through from time to time, but there are a lot of there here and you're already saying you'll make changes to the overall set of commits.

**Comment by Spevacus** [comment]
Yeah, some of them were 100% unintentional (merged in a branch that I didn't intend). I'll be certain to clean them up when I have a bit more time.

**Comment by Spevacus** [comment]
I've closed this in favor of #11549 since I wasn't able to do crazy rebase magic to make the commit history better here.
